### PR TITLE
Added bytes usage for Custom Partitioner

### PR DESCRIPTION
**Pull request by chaserileyroberts**
Added the usage of `bytes` for the backend_string key in `custom_partitioning.py`

This is a forwards-compatible fix for https://github.com/openxla/xla/pull/6719

### CODE REVIEW

**Review by pschuh — APPROVED** [review]

File: jax/experimental/custom_partitioning.py [hunk 1]
```diff
@@ -503,7 +506,7 @@ def to_mesh_pspec_sharding(hlo_sharding: Optional[xc.HloSharding], ndim):
   sharding_callback_info = _ShardingCallbackInfo(propagate_user_sharding,
       partition, to_mesh_pspec_sharding, in_tree, out_tree,
       infer_sharding_from_operands, ctx.module_context, mesh, static_args)
-  key = str(id(sharding_callback_info))
+  key = bytes(str(id(sharding_callback_info)))
```
Comment (pschuh) [inline jax/experimental/custom_partitioning.py:509]: bytes(..., "utf8")

Comment (chaserileyroberts) [reply jax/experimental/custom_partitioning.py]: Done.

**Comment by pschuh** [comment]
All merged. https://github.com/openxla/xla/pull/6719#issuecomment-1805263318

**Comment by pschuh** [comment]
I merged this with some changes, so it didn't get closed by copybara. Sorry about the trouble.

**Pull request closed without merging**